File: common/src/main/java/generations/gg/generations/core/generationscore/world/item/tools/GenerationsTiers.java
package generations.gg.generations.core.generationscore.world.item.tools;

import com.cobblemon.mod.common.CobblemonItems;
import com.google.common.base.Suppliers;
import generations.gg.generations.core.generationscore.world.item.GenerationsItems;
import generations.gg.generations.core.generationscore.world.level.block.GenerationsBlocks;
import net.minecraft.tags.BlockTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.Tier;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.level.block.Block;
import org.jetbrains.annotations.NotNull;

import java.util.function.Supplier;

public enum GenerationsTiers implements Tier {
    CHARGE_STONE(1, 185, 5.0F, 1.2F, 8, () -> GenerationsBlocks.CHARGE_STONE_SET.getBaseBlock().asItem()),
    VOLCANIC_STONE(1, 185, 5.0F, 1.2F, 8, () -> GenerationsBlocks.VOLCANIC_STONE.get().asItem()),
    AMETHYST(2, 300, 6.5F, 2.0F, 14, () -> Items.AMETHYST_SHARD),
    CRYSTAL(2, 300, 6.5F, 2.0F, 14, GenerationsItems.CRYSTAL),
    DAWN_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.DAWN_STONE),
    DUSK_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.DUSK_STONE),
    FIRE_STONE(3, 1561, 8.0F, 3.0F, 10, () -> CobblemonItems.FIRE_STONE),
    ICE_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.ICE_STONE),
    LEAF_STONE(2, 250, 6.0F, 2.0F, 14, () -> CobblemonItems.LEAF_STONE),
    MOON_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.MOON_STONE),

    RUBY(2, 300, 6.5F, 2.0F, 14, GenerationsItems.RUBY),

    SAPPHIRE(2, 300, 6.5F, 2.0F, 14, GenerationsItems.SAPPHIRE),
    SILICON(2, 100, 45F, 10F, 30, GenerationsItems.SILICON),
    SUN_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.SUN_STONE),
    THUNDER_STONE(3, 1561, 12.0F, 0.0F, 22, () -> CobblemonItems.THUNDER_STONE),
    WATER_STONE(3, 1561, 8.0F, 3.0F, 10, () -> CobblemonItems.WATER_STONE);

    private final int level;
    private final int uses;
    private final float speed;
    private final float damage;
    private final int enchantmentValue;
    private final com.google.common.base.Supplier<Ingredient> repairIngredient;

    GenerationsTiers(int level, int uses, float speed, float damage, int enchantmentValue, Supplier<Item> repairIngredient) {
        this.level = level;
        this.uses = uses;
        this.speed = speed;
        this.damage = damage;
        this.enchantmentValue = enchantmentValue;
        this.repairIngredient = Suppliers.memoize(() -> Ingredient.of(repairIngredient.get()));
    }

    public int getUses() {
        return this.uses;
    }

    public float getSpeed() {
        return this.speed;
    }

    public float getAttackDamageBonus() {
        return this.damage;
    }

    public int getLevel() {
        return this.level;
    }

    public int getEnchantmentValue() {
        return this.enchantmentValue;
    }

    public @NotNull Ingredient getRepairIngredient() {
        return this.repairIngredient.get();
    }

    public TagKey<Block> getTag() {
        return switch (this) {
            case CHARGE_STONE, VOLCANIC_STONE -> BlockTags.NEEDS_STONE_TOOL;
            case AMETHYST, CRYSTAL, LEAF_STONE, RUBY, SAPPHIRE, SILICON -> BlockTags.NEEDS_IRON_TOOL;
            case DAWN_STONE, DUSK_STONE, FIRE_STONE, ICE_STONE, MOON_STONE, SUN_STONE, THUNDER_STONE, WATER_STONE ->
                    BlockTags.NEEDS_DIAMOND_TOOL;
        };
    }

    @Override
    public String toString() {
        return super.toString();
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/world/item/tools/GenerationsTiers.kt
package generations.gg.generations.core.generationscore.common.world.item.tools

import com.cobblemon.mod.common.CobblemonItems
import com.google.common.base.Suppliers
import generations.gg.generations.core.generationscore.common.tags.GenerationsBlockTags
import generations.gg.generations.core.generationscore.common.world.item.GenerationsItems
import generations.gg.generations.core.generationscore.common.world.level.block.GenerationsBlocks
import net.minecraft.tags.BlockTags
import net.minecraft.tags.TagKey
import net.minecraft.world.item.Items
import net.minecraft.world.item.Tier
import net.minecraft.world.item.crafting.Ingredient
import net.minecraft.world.level.block.Block

enum class GenerationsTiers(
    private val incorrectBlocksForDrops: TagKey<Block>,
    val level: Int,
    private val uses: Int,
    private val speed: Float,
    private val damage: Float,
    private val enchantmentValue: Int,
    ingredientSupplier: () -> Ingredient
) : Tier {
    CHARGE_STONE(GenerationsBlockTags.INCORRECT_FOR_CHARGE_STONE_TOOL, 1, 131, 5.0f, 1.0f, 5, { Ingredient.of(GenerationsBlocks.CHARGE_STONE_SET.baseBlock) }),
    VOLCANIC_STONE(GenerationsBlockTags.INCORRECT_FOR_VOLCANIC_STONE_TOOL, 1, 131, 4.0f, 1.0f, 5, { Ingredient.of(GenerationsBlocks.VOLCANIC_STONE.value()) }),

    AMETHYST(GenerationsBlockTags.INCORRECT_FOR_AMETHYST_TOOL, 2, 250, 6f, 2f, 14, { Ingredient.of(Items.AMETHYST_SHARD) }),
    CRYSTAL(GenerationsBlockTags.INCORRECT_FOR_CRYSTAL_TOOL, 2, 250, 6f, 2f, 14, { Ingredient.of(GenerationsItems.CRYSTAL.value()) }),
    SAPPHIRE(GenerationsBlockTags.INCORRECT_FOR_SAPPHIRE_TOOL, 2, 250, 6f, 2f, 14, { Ingredient.of(GenerationsItems.SAPPHIRE.value()) }),
    SILICON(GenerationsBlockTags.INCORRECT_FOR_SILICON_TOOL, 2, 250, 6f, 2f, 14, { Ingredient.of(GenerationsItems.SILICON.value()) }),
    RUBY(GenerationsBlockTags.INCORRECT_FOR_RUBY_TOOL, 2, 250, 6f, 2f, 14, { Ingredient.of(GenerationsItems.RUBY.value()) }),

    DAWN_STONE(GenerationsBlockTags.INCORRECT_FOR_DAWN_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.DAWN_STONE) }),
    DUSK_STONE(GenerationsBlockTags.INCORRECT_FOR_DUSK_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.DUSK_STONE) }),
    FIRE_STONE(GenerationsBlockTags.INCORRECT_FOR_FIRE_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.FIRE_STONE) }),
    ICE_STONE(GenerationsBlockTags.INCORRECT_FOR_ICE_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.ICE_STONE) }),
    LEAF_STONE(GenerationsBlockTags.INCORRECT_FOR_LEAF_STONE_TOOL, 2, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.LEAF_STONE) }),
    MOON_STONE(GenerationsBlockTags.INCORRECT_FOR_MOON_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.MOON_STONE) }),
    SUN_STONE(GenerationsBlockTags.INCORRECT_FOR_SUN_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.SUN_STONE) }),
    THUNDER_STONE(GenerationsBlockTags.INCORRECT_FOR_THUNDER_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.THUNDER_STONE) }),
    WATER_STONE(GenerationsBlockTags.INCORRECT_FOR_WATER_STONE_TOOL, 3, 1561, 12f, 3f, 10, { Ingredient.of(CobblemonItems.WATER_STONE) }),
    ULTRITE(GenerationsBlockTags.INCORRECT_FOR_ULTRITE_TOOL, 5, 2640, 10.0f, 5.0f, 20, { Ingredient.of(GenerationsItems.ULTRITE_INGOT.value()) });

    private val repairIngredient = Suppliers.memoize(ingredientSupplier)

    override fun getUses(): Int {
        return this.uses
    }

    override fun getSpeed(): Float {
        return this.speed
    }

    override fun getAttackDamageBonus(): Float {
        return this.damage
    }

    override fun getIncorrectBlocksForDrops(): TagKey<Block> {
        return incorrectBlocksForDrops
    }

    override fun getEnchantmentValue(): Int {
        return this.enchantmentValue
    }

    override fun getRepairIngredient(): Ingredient = repairIngredient.get()

    val tag: TagKey<Block>
        get() = when (this) {
            CHARGE_STONE, VOLCANIC_STONE -> BlockTags.NEEDS_STONE_TOOL
            AMETHYST, CRYSTAL, LEAF_STONE, RUBY, SAPPHIRE, SILICON -> BlockTags.NEEDS_IRON_TOOL
            DAWN_STONE, DUSK_STONE, FIRE_STONE, ICE_STONE, MOON_STONE, SUN_STONE, THUNDER_STONE, WATER_STONE, ULTRITE -> BlockTags.NEEDS_DIAMOND_TOOL
        }

    override fun toString(): String {
        return super.toString()
    }
}
